### PR TITLE
337: Accordion Styling

### LINKED ISSUE #337: Implement "Accordion" Block
Implement a accordion block, based on boilerplate https://www.aem.live/developer/block-collection/accordion which is used on several pages from Insights.

Example:

https://www.sap.com/insights/data-management-glossary.html
https://main--hlx-test--urfuwo.hlx.page/resources/data-management-glossary

- [x] Basic Block implementation
- [ ] Final Styling

### PR DESCRIPTION
**Pull request by rjwtrmn**
Adding styling for accordion block to match https://sap.frontify.com/document/223143#/-/accordion

There is potentially some additional functionality to be added wrt a container header and "Expand/Collapse All" button. Here we've just focused on styling/behaviour in the body of the accordion.

Fix #337 

Test URLs:
- Before: https://main--hlx-test--urfuwo.hlx.live/resources/data-management-glossary
- After: https://337-accordion-styling--hlx-test--urfuwo.hlx.live/resources/data-management-glossary

### CODE REVIEW

**Comment by saurabh-khare** [comment]
@rjwtrmn Is this PR up for review or draft? If its ready for review then please add reviewers.

**Review by mhaack — APPROVED** [review]

**Review by rootpea — APPROVED** [review]

**Comment by saurabh-khare** [comment]
@rjwtrmn Can accordion only have text? What if we put image?

**Comment by saurabh-khare** [comment]
Please add a transition for rotating arrow and accordion slide

File: aemedge/blocks/accordion/accordion.css [hunk 1]
```diff
@@ -1,57 +1,48 @@
...
-  position: absolute;
-  top: 50%;
-  right: 18px;
-  transform: translateY(-50%) rotate(135deg);
-  width: 9px;
-  height: 9px;
-  border: 2px solid;
-  border-width: 2px 2px 0 0;
-  transition: transform 0.2s;
+.accordion details[open] .accordion-arrow {
```
Comment (saurabh-khare) [inline aemedge/blocks/accordion/accordion.css:38]: Can this be changed from `open` to `aria-expanded`?

File: aemedge/blocks/accordion/accordion.css [hunk 2]
```diff
@@ -1,57 +1,48 @@
...
-}
+  /* Helps ensure icon and text are aligned properly */
+  line-height: var(--udexTypographyHeadingMediumXXXSFontSize);
+  width: 42px;
+  height: 42px;
 
-.accordion details summary::-webkit-details-marker {
-  display: none;
+  /* Icon image should be 18px x 18px */
+  padding: calc((42px - 18px) / 2);
```
Comment (saurabh-khare) [inline aemedge/blocks/accordion/accordion.css:34]: Why require `calc` function if values are fixed?

File: aemedge/blocks/accordion/accordion.css [hunk 3]
```diff
@@ -1,57 +1,48 @@
...
-  border: 2px solid;
-  border-width: 2px 2px 0 0;
-  transition: transform 0.2s;
+.accordion details[open] .accordion-arrow {
+  transform: rotate(90deg);
 }
 
-.accordion details[open] summary::after {
-  transform: translateY(-50%) rotate(-45deg);
+.accordion details summary::-webkit-details-marker {
```
Comment (saurabh-khare) [inline aemedge/blocks/accordion/accordion.css:42]: Why require a vendor prefix pseudo element?

File: aemedge/blocks/accordion/accordion.js [hunk 4]
```diff
@@ -12,22 +15,26 @@ export default function decorate(block) {
...
-    const details = document.createElement('details');
-    details.className = 'accordion-item';
-    details.append(summary, body);
+    const details = domEl(
+      'details',
+      { class: 'accordion-item' },
+      summary,
+      body,
+    );
+    decorateIcons(details);
```
Comment (saurabh-khare) [inline aemedge/blocks/accordion/accordion.js:37]: Can we run it outside the loop? May be call `decorateIcons(block)`? It should be able to decorate all icons inside

**Review by saurabh-khare — CHANGES_REQUESTED** [review]

File: aemedge/blocks/accordion/accordion.css (hunk 3 above)
Comment (rjwtrmn) [reply aemedge/blocks/accordion/accordion.css:42]: This was part of the styling provided in the base implementation of accordion, I assume this does something to hide some webkit accordion browser element and didn't want to remove.

**Comment by rjwtrmn** [comment]
> @rjwtrmn Can accordion only have text? What if we put image?

I haven't seen any examples of this anywhere, only text. As far as I am aware we will only have text in the accordion for now.

File: aemedge/blocks/accordion/accordion.css (hunk 3 above)
Comment (saurabh-khare) [reply aemedge/blocks/accordion/accordion.css:42]: Check if it does nothing then remove it. 
@mhaack It is discouraged to use vendor prefixes. Can we check why it is in reference code?

**Comment by saurabh-khare** [comment]
> > Please add a transition for rotating arrow and accordion slide
> 
> @saurabh-khare Do we just make up our own values for this or is there any guidance available for us regarding animations?

Yes general guidance is available here:
https://sap.frontify.com/document/223141#/movement/interaction/css-transitions

**Comment by saurabh-khare** [comment]
> > @rjwtrmn Can accordion only have text? What if we put image?
> 
> @saurabh-khare I haven't seen any examples of this anywhere, only text. As far as I am aware we will only have text in the accordion for now.

@mhaack Can you confirm? Or should we ask from Elaine?

File: aemedge/blocks/accordion/accordion.css (hunk 1 above)
Comment (rjwtrmn) [reply aemedge/blocks/accordion/accordion.css:38]: `details` elements seem to have an implicit aria role of "group", which isn't included in the list of [roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded#associated_roles) that use `aria-expanded`. I guess we should stick with `open`.

File: aemedge/blocks/accordion/accordion.css (hunk 3 above)
Comment (mhaack) [reply aemedge/blocks/accordion/accordion.css:42]: Lets not complicate it, se https://pages.github.tools.sap/UDEx/digital-design-system/?path=/docs/patterns-actions-accordion--docs that also has no transition. Let's keep it simpler.

> It is discouraged to use vendor prefixes
If this is for some styles only I would be fine. I that impacts functionality we should not use it.

File: aemedge/blocks/accordion/accordion.js (hunk 4 above)
Comment (mhaack) [reply aemedge/blocks/accordion/accordion.js:37]: +1

**Comment by mhaack** [comment]
> > > @rjwtrmn Can accordion only have text? What if we put image?
> > 
> > 
> > @saurabh-khare I haven't seen any examples of this anywhere, only text. As far as I am aware we will only have text in the accordion for now.
> 
> @mhaack Can you confirm? Or should we ask from Elaine?

The samples I have seen so far only had text and links only. However I agree everything an author puts into the table cell (images, lists, links) should work.

**Comment by rjwtrmn** [comment]
> > > > @rjwtrmn Can accordion only have text? What if we put image?
> > > 
> > > 
> > > @saurabh-khare I haven't seen any examples of this anywhere, only text. As far as I am aware we will only have text in the accordion for now.
> > 
> > 
> > @mhaack Can you confirm? Or should we ask from Elaine?
> 
> The samples I have seen so far only had text and links only. However I agree everything an author puts into the table cell (images, lists, links) should work.

I have added an example with an image to the blocks library. We are applying outer padding but no other styles. Surely we would need some sort of requirements for styling these other cases? We can adjust the stylesheet as the need arises.

File: aemedge/blocks/accordion/accordion.css (hunk 3 above)
Comment (saurabh-khare) [reply aemedge/blocks/accordion/accordion.css]: > Lets not complicate it, se https://pages.github.tools.sap/UDEx/digital-design-system/?path=/docs/patterns-actions-accordion--docs that also has no transition. Let's keep it simpler.
> 
> > It is discouraged to use vendor prefixes
> > If this is for some styles only I would be fine. I that impacts functionality we should not use it.

Sure. On a side note I also see an "Expand All" button in their implementation. Should we also add that?
Screenshot 2024-03-26 at 14 47 04

Comment (rjwtrmn) [reply aemedge/blocks/accordion/accordion.css]: I have removed the prefix as it nothing on Chrome or Firefox. Perhaps someone with a Mac can check on Safari and see if we should add it back, @saurabh-khare would you be able to confirm?

Comment (rjwtrmn) [reply aemedge/blocks/accordion/accordion.css]: > > Lets not complicate it, se https://pages.github.tools.sap/UDEx/digital-design-system/?path=/docs/patterns-actions-accordion--docs that also has no transition. Let's keep it simpler.
> > > It is discouraged to use vendor prefixes
> > > If this is for some styles only I would be fine. I that impacts functionality we should not use it.
> 
> Sure. On a side note I also see an "Expand All" button in their implementation. Should we also add that? Screenshot 2024-03-26 at 14 47 04

There is a follow up ticket raised for this https://github.com/urfuwo/hlx-test/issues/385

Comment (saurabh-khare) [reply aemedge/blocks/accordion/accordion.css]: > I have removed the prefix as it did nothing on Chrome or Firefox. Perhaps someone with a Mac can check on Chrome/Firefox/Safari and see if we should add it back, @saurabh-khare would you be able to confirm?

Arrghh I see ugly arrows in Safari. Add it back. I'll see later how can we avoid it.
Screenshot 2024-03-26 at 15 08 31

**Review by saurabh-khare — APPROVED** [review]

**Review by mhaack — APPROVED** [review]

**Pull request merged**